File: src/go/bloomfilter/bloomfilter_test.go
package bloomfilter

import (
	"testing"
)

func TestBloomFilterSanity(t *testing.T) {
	bf, err := Create(0.1, 100)
	if err != nil {
		t.Errorf("Failed to create a bloom filter %v", err)
	}
	if bf == nil {
		t.Errorf("Failed to create a bloom filter: nil")
	}
	defer bf.Destroy()

	if bf.NumHashFunctions() == 0 {
		t.Errorf("invalid bloom filter: nil")
	}
}

func TestBloomFilterFailure(t *testing.T) {
	bf := BloomFilter{}

	err := bf.Add("a")
	if err == nil {
		t.Errorf("adding values to an invalid context should fail")
	}

	err = bf.AddArray([]string{"a", "b"})
	if err == nil {
		t.Errorf("adding values to an invalid context should fail")
	}

	_, err = bf.Check("a")
	if err == nil {
		t.Errorf("checking values in an invalid context should fail")
	}

	_, err = bf.ToJSON()
	if err == nil {
		t.Errorf("create json should fail with an invalid context")
	}

	err = bf.Destroy()
	if err == nil {
		t.Errorf("destroy should fail with an invalid context")
	}

	cnt := bf.NumHashFunctions()
	if cnt != 0 {
		t.Errorf("NumHashFunctions should fail with an invalid context")
	}
	_, err = Create(3, 100)
	if err == nil {
		t.Errorf("Create should fail with an invalid fpr")
	}

	_, err = CreateFromJSON("abc")
	if err == nil {
		t.Errorf("CreateFromJSON should fail with an invalid input")
	}
}

func TestBloomFilterAdd(t *testing.T) {
	bf, err := Create(0.001, 1<<10)
	if err != nil || bf == nil {
		t.Errorf("Failed to create a bloom filter %v", err)
	}
	defer bf.Destroy()

	err = bf.Add("a")
	if err != nil {
		t.Errorf("Failed to add element to the bloom filter %v", err)
	}

	valid, err := bf.Check("a")
	if valid == false || err != nil {
		t.Errorf("Failed to add element to bloom filter")
	}
	invalid, err := bf.Check("b")
	if invalid == true || err != nil {
		t.Errorf("bloom filter check fail")
	}
}

func TestBloomFilterAddArray(t *testing.T) {
	bf, err := Create(0.001, 1<<10)
	if err != nil || bf == nil {
		t.Errorf("Failed to create a bloom filter %v", err)
	}
	defer bf.Destroy()

	err = bf.AddArray([]string{"a", "b", "c", "d"})
	if err != nil {
		t.Errorf("Failed to add array to the bloom filter %v", err)
	}

	for _, val := range []string{"a", "b", "b", "d"} {
		valid, err := bf.Check(val)
		if valid == false || err != nil {
			t.Errorf("Failed to add element to bloom filter")
		}
	}
	invalid, err := bf.Check("z")
	if invalid == true || err != nil {
		t.Errorf("bloom filter check fail")
	}
}

func TestBloomFilterFPR(t *testing.T) {
	for maxElements := 1 << 10; maxElements < (1 << 20); maxElements *= 2 {
		targetFPR := 0.1
		bf, err := Create(targetFPR, int64(maxElements))
		if err != nil || bf == nil {
			t.Errorf("Failed to create a bloom filter %v", err)
		}

		for i := 0; i < maxElements; i++ {
			bf.Add("Element " + string(i))
		}

		count := 0
		numTests := 10000
		for i := 0; i < numTests; i++ {
			found, _ := bf.Check("Test " + string(i))
			if found {
				count++
			}
		}

		actualFPR := float64(count / numTests)
		if actualFPR > (1.2 * targetFPR) {
			t.Errorf("FPR test fail. actual %v expected %v", actualFPR, 1.2*targetFPR)
		}
		bf.Destroy()
	}
}

func TestBloomFilterToJSON(t *testing.T) {
	bf, err := Create(0.01, 100)
	if err != nil || bf == nil {
		t.Errorf("Failed to create a bloom filter %v", err)
	}
	defer bf.Destroy()

	for i := 0; i < 100; i++ {
		bf.Add("Element " + string(i))
	}
	encoded, err := bf.ToJSON()
	if err != nil {
		t.Errorf("Failed to create json %v", err)
	}
	decoded, err := CreateFromJSON(encoded)
	if err != nil || decoded == nil {
		t.Errorf("Failed to decode bloom filter %v", err)
	}
	defer decoded.Destroy()

	for i := 0; i < 100; i++ {
		found, err := decoded.Check("Element " + string(i))
		if found == false || err != nil {
			t.Errorf("invalid BF decoded")
		}
	}
}

File: psi_cardinality/cpp/bloom_filter_test.cpp
//
// Copyright 2020 the authors listed in CONTRIBUTORS.md
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include "bloom_filter.h"

#include "absl/strings/str_cat.h"
#include "gtest/gtest.h"
#include "util/status_matchers.h"

namespace psi_cardinality {
namespace {

class BloomFilterTest : public ::testing::Test {
 protected:
  void SetUp() { return SetUp(0.001, 1 << 10); }
  void SetUp(double fpr, int max_elements) {
    PSI_ASSERT_OK_AND_ASSIGN(filter_, BloomFilter::Create(fpr, max_elements));
  }

  std::unique_ptr<BloomFilter> filter_;
};

TEST_F(BloomFilterTest, TestAdd) {
  std::vector<std::string> elements = {"a", "b", "c", "d"};

  // Test both variants of Add.
  filter_->Add(elements[0]);
  filter_->Add(absl::MakeConstSpan(&elements[1], elements.size() - 1));

  // Check if all elements are present.
  for (const auto& element : elements) {
    EXPECT_TRUE(filter_->Check(element));
  }
  EXPECT_FALSE(filter_->Check("not present"));
}

TEST_F(BloomFilterTest, TestFPR) {
  for (int max_elements = 1 << 10; max_elements < (1 << 20);
       max_elements *= 2) {
    double target_fpr = 0.1;
    SetUp(target_fpr, max_elements);
    // Insert `max_elements` elements.
    for (int i = 0; i < max_elements; i++) {
      filter_->Add(absl::StrCat("Element ", i));
    }
    // Test 10k elements to measure FPR.
    double count = 0;
    int num_tests = 10000;
    for (int i = 0; i < num_tests; i++) {
      if (filter_->Check(absl::StrCat("Test ", i))) {
        count++;
      }
    }
    // Check if actual FPR matches the target FPR, allowing for 20% error.
    double actual_fpr = count / num_tests;
    EXPECT_LT(actual_fpr, 1.2 * target_fpr)
        << absl::StrCat("max_elements: ", max_elements);
  }
}

TEST_F(BloomFilterTest, TestToJSON) {
  double fpr = 0.01;
  int max_elements = 100;
  SetUp(fpr, max_elements);
  for (int i = 0; i < max_elements; i++) {
    filter_->Add(absl::StrCat("Element ", i));
  }

  // Encode Bloom filter as JSON and check if it matches.
  std::string encoded_filter = filter_->ToJSON();
  std::string expected =
      "{\"num_hash_functions\":7,\"bits\":\"VN3/"
      "BXfUjEDvJLcxCTepUCTXGQwlTax0xHiMohCNb45uShFsznK099RH0CFVIMn91Bdc7jLkXH"
      "Xr"
      "Xp1NimmZSDrYSj5sd/"
      "500nroNOdXbtd53u8cejPMGxbx7kR1E1zyO19mSkYLXq4xf7au5dFN0qhxqfLnjaCE\"}";
  EXPECT_EQ(encoded_filter, expected);
}

TEST_F(BloomFilterTest, TestCreateFromJSON) {
  std::vector<std::string> elements = {"a", "b", "c", "d"};
  filter_->Add(elements);
  std::string encoded_filter = filter_->ToJSON();
  PSI_ASSERT_OK_AND_ASSIGN(auto filter2,
                           BloomFilter::CreateFromJSON(encoded_filter));
  for (const auto& element : elements) {
    EXPECT_TRUE(filter2->Check(element));
  }
  EXPECT_FALSE(filter2->Check("not present"));
}

}  // namespace
}  // namespace psi_cardinality